packages/astro/src/env/vite-plugin-env-virtual-mod.ts

```diff
@@ -0,0 +1,226 @@
+import { loadEnv, type Plugin } from 'vite';
+import type { AstroSettings } from '../@types/astro.js';
+import type { Logger } from '../core/logger/core.js';
+import {
+	ENV_TYPES_FILE,
+	MODULE_TEMPLATE_URL,
+	TYPES_TEMPLATE_URL,
+	VIRTUAL_MODULES_IDS,
+} from './constants.js';
+import type { EnvSchema } from './schema.js';
+import { getType, validateEnvVariable } from './validators.js';
+import { AstroError, AstroErrorData } from '../core/errors/index.js';
+import type fsMod from 'node:fs';
+import { fileURLToPath } from 'node:url';
+
+// TODO: reminders for when astro:env comes out of experimental
+// Types should always be generated (like in types/content.d.ts). That means the client module will be empty
+// and server will only contain getSecret for unknown variables. Then, specifying a schema should only add
+// variables as needed. For secret variables, it will only require specifying SecretValues and it should get
+// merged with the static types/content.d.ts
+
+interface AstroEnvVirtualModPluginParams {
+	settings: AstroSettings;
+	logger: Logger;
+	mode: 'dev' | 'build' | string;
+	fs: typeof fsMod;
+}
+
+export function astroEnvVirtualModPlugin({
+	settings,
+	logger,
+	mode,
+	fs,
+}: AstroEnvVirtualModPluginParams): Plugin | undefined {
+	if (!settings.config.experimental.env) {
+		return;
+	}
+	logger.warn('env', 'This feature is experimental.');
```

Comment:
Do we need a warning here? If they opt into it in the config, I think it should be enough that they know it's experimental.